Source/items.cpp

```diff
@@ -3331,32 +3336,23 @@ void TryRandomUniqueItem(Item &item, _item_indexes idx, int8_t mLevel, int uper,
 		return;
 	}
 
-	const UniqueItem &uniqueItem = UniqueItems[uid];
-	int targetLvl = 1; // Target level for reverse compatibility, since vanilla always takes the last applicable uid in the list.
-
-	// Set target level. Ideally we use uper 15 to have a 16% chance of generating a unique item.
-	if (uniqueItem.UIMinLvl - 4 > 0) { // Negative level will underflow. Lvl 0 items may have unintended consequences.
-		uper = 15;
-		targetLvl = uniqueItem.UIMinLvl - 4;
-	} else {
-		uper = 1;
-		targetLvl = uniqueItem.UIMinLvl;
+	// Find our own id to calculate the offset in validUniques
+	int uidOffset = -1;
+	for (size_t i = 0; i < validUniques.size(); i++) {
+		if (validUniques[i] == uid) {
+			// vanilla always picks the last unique, so the offset is calculated from the back of the valid unique list
+			uidOffset = static_cast<int>(validUniques.size() - i - 1);
+		}
 	}
+	assert(uidOffset != -1);
 
-	// Amount to decrease the final uid by in CheckUnique() to get the desired unique.
-	const auto uidOffset = static_cast<int>(std::count_if(UniqueItems.begin() + uid + 1, UniqueItems.end(), [&uniqueItem](UniqueItem &potentialMatch) {
-		return uniqueItem.UIItemId == potentialMatch.UIItemId && uniqueItem.UIMinLvl == potentialMatch.UIMinLvl;

```

Comment:
`return uniqueItem.UIItemId == potentialMatch.UIItemId && uniqueItem.UIMinLvl >= potentialMatch.UIMinLvl;`
Infinite loop fix

Reply:
Not entirely, see stephan's comment in your pr. also I think it's nice if we use the same logic to generate the valid uniques as vanilla did (same item level and reuse logic to avoid different result).